src/dbnode/storage/index/postings_list_cache.go

```diff
@@ -0,0 +1,253 @@
+	patternType PatternType,
+) (postings.List, bool) {
+	// No RLock because a Get() operation mutates the LRU.
+	q.Lock()
+	p, ok := q.lru.Get(segmentUUID, pattern, patternType)
+	q.Unlock()
+
+	q.emitCacheGetMetrics(patternType, ok)
+
+	if !ok {
+		return nil, false
+	}
+
+	return p, ok
+}
+
+// PutRegexp updates the LRU with the result of the regexp query.
+func (q *PostingsListCache) PutRegexp(
+	segmentUUID uuid.UUID,
+	pattern string,
+	pl postings.List,
+) {
+	q.put(segmentUUID, pattern, PatternTypeRegexp, pl)
+}
+
+// PutTerm updates the LRU with the result of the term query.
+func (q *PostingsListCache) PutTerm(
+	segmentUUID uuid.UUID,
+	pattern string,
+	pl postings.List,
+) {
+	q.put(segmentUUID, pattern, PatternTypeTerm, pl)
+}
+
+func (q *PostingsListCache) put(
+	segmentUUID uuid.UUID,
+	pattern string,
+	patternType PatternType,
+	pl postings.List,
+) {
+	q.Lock()
+	q.lru.Add(
+		segmentUUID,
+		pattern,
+		patternType,
+		pl,
+	)
+	q.Unlock()
+	q.emitCachePutMetrics(patternType)
+}
+
+// PurgeSegment removes all postings lists associated with the specified
+// segment from the cache.
+func (q *PostingsListCache) PurgeSegment(segmentUUID uuid.UUID) {
+	q.Lock()
+	q.lru.PurgeSegment(segmentUUID)
+	q.Unlock()
+}
+
+// StartReportLoop starts a background process that will call Report()
+// on a regular basis and returns a function that will end the background
+// process.
+func (q *PostingsListCache) StartReportLoop() func() {
+	doneCh := make(chan struct{})
+
+	go func() {
+		for {
+			select {
+			case <-doneCh:
+				return
+			default:
+			}
+
+			q.Report()
+			time.Sleep(reportLoopInterval)
+		}
+	}()
+
+	return func() { close(doneCh) }
+}
+
+// Report will emit metrics about the status of the cache.
+func (q *PostingsListCache) Report() {
+	var (
+		size     float64
+		capacity float64
+	)
+
+	q.Lock()
+	size = float64(q.lru.Len())
+	capacity = float64(q.size)
+	q.Unlock()
+
+	q.metrics.size.Update(size)
+	q.metrics.capacity.Update(capacity)
+}
+
+func (q *PostingsListCache) emitCacheGetMetrics(patternType PatternType, hit bool) {
+	switch {
+	case patternType == PatternTypeRegexp && hit:
+		q.metrics.regexp.hits.Inc(1)
+	case patternType == PatternTypeRegexp && !hit:
+		q.metrics.regexp.misses.Inc(1)
+	case patternType == PatternTypeTerm && hit:
+		q.metrics.term.hits.Inc(1)
+	case patternType == PatternTypeTerm && !hit:
+		q.metrics.term.misses.Inc(1)
+	}
+}
+
+func (q *PostingsListCache) emitCachePutMetrics(patternType PatternType) {
+	if patternType == PatternTypeRegexp {
+		q.metrics.regexp.puts.Inc(1)
+	} else {
+		q.metrics.term.puts.Inc(1)
+	}
+}
+
+type postingsListCacheMetrics struct {
+	regexp *postingsListCacheMethodMetrics
+	term   *postingsListCacheMethodMetrics
+
+	size     tally.Gauge
+	capacity tally.Gauge
+}
+
+func newPostingsListCacheMetrics(scope tally.Scope) *postingsListCacheMetrics {
+	return &postingsListCacheMetrics{
+		regexp: newPostingsListCacheMethodMetrics(scope.SubScope("regexp")),
```

Comment:
probably better to have these as tags instead of subscopes. would allow for querying with the m3ql functions better

Reply:
still applicable?

Reply:
sure thing